enhancements/kube-apiserver/static-pods-for-single-node-via-ignition.md

```diff
@@ -0,0 +1,92 @@
+
+## Open Questions [optional]
+
+1. Do we even have a use-case for spending the time building this thought experiment?
+2. Do we want to be able to run other operators on top of this?
+3. Do we want things like must-gather to function properly?
+
+## Summary
+
+A while back, Seth Jennings had a cool idea for trying to create a single node cluster using ignition.
+The cluster would be non-configurable after "creation", non-upgradable, non-HA.
+The cluster would only have etcd, kube-apiserver, kube-controller-manager, kube-scheduler.
+This is a description of how we could generate supportable static pods.
+
+## Motivation
+
+Documenting a thought experiment about single node clusters.
+
+### Goals
+
+1. Create a supportable static pod.
+   Ideally, this looks very similar to our existing static pods.
+2. Some amount of configuration is important, how flexible can we be?
+
+### Non-Goals
+
+1. Try to run a kube control plane operator.
+   This is unopinionated about what goes on top, but the kube control plane will not be reconfigured after the fact.
+
+## Proposal
+
+We can create a new kind of render command which takes existing inputs *and* config.openshift.io resources.
+Similar to how we built the original disaster recovery for certificates, we can factor the command to run the various
+control loops "in order".
+We can initialize our control loops using fake clients and wire listeners to synthetically update indexers backing fake
+listers.
+This is like we do for unit tests, only wired into the update reactors for the client.
+If we separate the reactive bits of the control loops, the informer watch triggers adn the like, from the data input bits
+(I think this is possible), we can have very high fidelity.
+In the kube-apiserver, the ordering would like this for instance:
+1.  cert-rotation - we need to create certs
+2.  encryption - this would need a special mode to say: just encrypt it right away
+3.  bound tokens - this creates some secrets for us
+4.  static-resources - this creates targets, SAs, and stuff
+5.  config observation - we need to set the operator observed config to be able to generate the final config.
+6.  target config - writes the kube-apiserver configmap
+7.  resource sync - copies bits from A to B
+8.  loop through config observation, target config, resource sync one more time (yeah, cycles)
+9.  revision controller
+
+Now we do a couple neat things:
+1. Export all content from the fake clients to produce resource manifests that will be created bootkube style against
+   the kube-apiserver.
+   Someone will have grown a dependency and we know for sure that the next operator will require input from the previous one.
+2. Wire up the fake clients to our installer command.
+   In theory, this command will create an exact copy of the "normal" kube-apiserver static pod that we create.
+
+This gives leaves supporting only one shape of static pods, which makes support of these static pods much easier.
```

Comment:
Is this in theory supportable so during a 4.y release cycle we could define static versions that people can templatize (minimally) with things like on disk certs?  I.e. would this "shape" to be roughly supportable with limited flexibility to change, without having to change the existing operator dramatically?

Reply:
I think our first attempt would be to get people to run this installer with an input that looks exactly like what they would use in a "real" cluster.  So we would accept a manifest containing their serving cert and a manifest containing their apiserver.config.openshift.io that says how to use it.

This would allow us...
1. to have one set of code managing the user input
2. having a single external interface to the world instead of promising the shape of a static pod
3. allow the cluster-admin to test/confirm his changes in a real cluster and take those settings as input to producing a single-node cluster.

If we start trying to allow injection of disk certs, our on-disk static pods become an API we need to support.